### PR TITLE
ARROW-4882: [GLib] Add sum functions

### PR DESCRIPTION
**Pull request by shiro615**

### CODE REVIEW

File: c_glib/arrow-glib/basic-array.cpp [hunk 1]
```diff
@@ -83,6 +83,15 @@ garrow_primitive_array_new(GArrowDataType *data_type,
...
 
+template <typename T>
+typename T::c_type
+garrow_numeric_array_get_sum_value_raw(std::shared_ptr<arrow::Scalar> scalar)
+{
+  using ScalarType = typename arrow::TypeTraits<T>::ScalarType;
+  auto arrow_numeric_scalar = std::dynamic_pointer_cast<ScalarType>(scalar);
+  return arrow_numeric_scalar->value;
+};
+
```
Comment (kou) [inline c_glib/arrow-glib/basic-array.cpp]: Can we create a template function that can be used by all `sum` functions?

```diff
diff --git a/c_glib/arrow-glib/basic-array.cpp b/c_glib/arrow-glib/basic-array.cpp
index deedc0fe..4db6a564 100644
--- a/c_glib/arrow-glib/basic-array.cpp
+++ b/c_glib/arrow-glib/basic-array.cpp
@@ -92,6 +92,28 @@ garrow_numeric_array_get_sum_value_raw(std::shared_ptr<arrow::Scalar> scalar)
   return arrow_numeric_scalar->value;
 };
 
+template <typename ArrowType, typename GArrowArrayType>
+typename ArrowType::c_type
+garrow_numeric_array_sum(GArrowArrayType array,
+                         GError **error,
+                         const gchar *tag,
+                         typename ArrowType::c_type default_value)
+{
+  auto arrow_array = garrow_array_get_raw(GARROW_ARRAY(array));
+  auto memory_pool = arrow::default_memory_pool();
+  arrow::compute::FunctionContext context(memory_pool);
+  arrow::compute::Datum sum_datum;
+  auto status = arrow::compute::Sum(&context, arrow_array, &sum_datum);
+  if (garrow_error_check(error, status, tag)) {
+    using ScalarType = typename arrow::TypeTraits<ArrowType>::ScalarType;
+    auto arrow_numeric_scalar =
+      std::dynamic_pointer_cast<ScalarType>(sum_datum.scalar());
+    return arrow_numeric_scalar->value;
+  } else {
+    return default_value;
+  }
+}
+
 G_BEGIN_DECLS
 
 /**
@@ -1062,16 +1084,10 @@ gint8
 garrow_int8_array_sum(GArrowInt8Array *array,
                       GError **error)
 {
-  auto arrow_array = garrow_array_get_raw(GARROW_ARRAY(array));
-  auto memory_pool = arrow::default_memory_pool();
-  arrow::compute::FunctionContext context(memory_pool);
-  arrow::compute::Datum sum_datum;
-  auto status = arrow::compute::Sum(&context, arrow_array, &sum_datum);
-  if (garrow_error_check(error, status, "[int8-array][sum]")) {
-    return garrow_numeric_array_get_sum_value_raw<arrow::Int64Type>(sum_datum.scalar());
-  } else {
-    return 0;
-  }
+  return garrow_numeric_array_sum<arrow::Int64Type>(array,
+                                                    error,
+                                                    "[int8-array][sum]",
+                                                    0);
 }
```

File: c_glib/arrow-glib/basic-array.cpp [hunk 2]
```diff
@@ -1039,6 +1048,32 @@ garrow_int8_array_get_values(GArrowInt8Array *array,
   return garrow_array_get_values_raw<arrow::Int8Type>(arrow_array, length);
 }
 
+/**
+ * garrow_int8_array_sum:
+ * @array: A #GArrowInt8Array.
+ * @error: (nullable): Return location for a #GError or %NULL.
+ *
+ * Returns: Returns: The value of the computed sum on success,
+ *   0 on error.
```
Comment (kou) [inline c_glib/arrow-glib/basic-array.cpp]: "0 on error" is misleading. 0 is one of a valid value. Users should not use 0 as an error indicator.

**Review by kou — CHANGES_REQUESTED** [review]
Thanks!
Could you check my comments?

File: c_glib/arrow-glib/basic-array.cpp (hunk 1 above)
Comment (shiro615) [reply c_glib/arrow-glib/basic-array.cpp]: Thanks! I've updated a template function.

File: c_glib/arrow-glib/basic-array.cpp [hunk 3]
```diff
@@ -1039,6 +1048,32 @@ garrow_int8_array_get_values(GArrowInt8Array *array,
...
+ *
+ * Since: 0.13.0
+ */
+gint8
```
[the remark this answers is not in the document]
Comment (shiro615) [reply c_glib/arrow-glib/basic-array.cpp]: Yes. I've fixed it.

File: c_glib/arrow-glib/basic-array.cpp (hunk 2 above)
Comment (shiro615) [reply c_glib/arrow-glib/basic-array.cpp]: I've fixed documents.

Comment (fsaintjacques) [reply c_glib/arrow-glib/basic-array.cpp]: Note that it may also return NULL.

Comment (kou) [reply c_glib/arrow-glib/basic-array.cpp]: Thanks.
We need to NULL check, then.

Comment (kou) [reply c_glib/arrow-glib/basic-array.cpp]: I've added.
We use zero for the case because we can't implement `option<T>` in C for primitive types.

**Review by kou — APPROVED** [review]
+1
I'll merge this when CI is passed.

File: c_glib/test/test-double-array.rb [hunk 4]
```diff
@@ -55,6 +55,6 @@ def test_values
 
   def test_sum
     array = build_float_array([1.5, 3.0, nil])
-    assert_equal(4.5, array.sum)
+    assert_in_delta(4.5, array.sum)
```
Comment (shiro615) [inline c_glib/test/test-double-array.rb]: Oh, sorry. I missed it.

**Comment by shiro615** [comment]
Thanks!

**Pull request closed without merging**